### ISSUE TITLE
Need an a way to allow Expression functionality: author fabric Code on the component level

### ISSUE THREAD

**Issue by bhaux** [fabric-engine/Kraken#453]
To fill the gap of missing expressions while rigging, it'd be great to create a non-preset fabric node on the fly within the component definition - to quickly set expression level computations between attributes etc. The implementation just needs to be lightweight - since most of them are going to be one-offs, such as a simple min/max function etc.

The most powerful and flexible solution could be just a call to a KLOperator - and instead of loading from a preset, expose the ability to set inputs, outputs and code string.

**Comment by EricTRocks** [comment]
Are you concerned about performance with these? MY thoughts on this would be that these operations would be bundled together with larger operators that take the inputs and calculate these values as higher level ops to ensure that there is not a tremendous amount of i/o with the Maya graph. Thoughts?

I'm sure we could implement this, just want to ensure that it's not going to drag down performance.

**Comment by bhaux** [comment]
I understand your concern but I'm not too worried about performance: This is not a problem we're currently seeing, as we are already creating these nodes - but as a uniquely created solver file for each and every Expression Calculation (ouch).  So we end up creating these klops anyway but it's a really backwards and inconvenient way. I would be more concerned if we had would push a huge amount of i/o's but this is mostly used for one-off expressions.

Not sure if I understand the second thought properly,  but as it is, we're already wrapping all computations which can and should be owned by larger KL operators/solvers into them - but both practically and conceptually this is a rare occurrence as each of those do a very specific thing, and we don't want to wrap any potential value pre/postprocessing into them unless absolutely necessary - to keep them lean, simple, reuseable, and avoid unnecessary bloating.

I would imagine this to be very simple and look e.g. like this:
`
self.klExpressionOp = KLExpression('RandomExpression')
self.addOperator(self.klExpressionOp)
# Add Inputs / Outputs
self.klExpressionOp.setInput('a', type=float, self.a_param)
self.klExpressionOp.setInput('b', type=float, self.b_param)
self.klExpressionOp.setOutput('result', self.result_param)
# Add Code
self.klExpressionOp.setCode('result =  (a + b) + max(1-b,0.5)')
`

**Comment by EricTRocks** [comment]
OK thanks for the feedback. I'll get working on this. It makes sense for sure I just wanted to raise the concern of performance but if you're not seeing any issue with performance currently we can get this done.

**Comment by todderasesareddot** [comment]
Yeah.  What Bernhard said.  We need our rig to be 100% Fabric in terms of every xfo and driven attribute so we need these smaller utilities.  We need the flexibility and I don't think we'd rack up too many of them.  This will be a nice addition. 

    On Monday, May 1, 2017 1:32 PM, Eric Thivierge <notifications@github.com> wrote:
 

 OK thanks for the feedback. I'll get working on this. It makes sense for sure I just wanted to raise the concern of performance but if you're not seeing any issue with performance currently we can get this done.—
